### DISCUSSION TITLE
Pytest vs Unittest framework

### DISCUSSION THREAD

**Discussion by guidopetri** [getredash/redash#6412] [Polls]
We have a sort of mix between `pytest` and `unittest` being used in our tests. This isn't ideal (it's a little confusing with all the mocking/patching) and is pretty confusing to a newcomer.

The unittest library is a built-in, but is more verbose and class-based / object-based.

The pytest library is an external dependency (though we already use it), and is more straightforward / function-based.

**Comment by guidopetri** [comment]
@getredash/maintainers , curious to hear your thoughts in particular.

**Comment by arikfr** [comment]
Some background: 

- We started with unittest. Probably because it's builtin, but don't really remember if there was a reason beyond that.
- Then we started using pytest in the SaaS codebase. Found it to be really nice and with potential to simplify our tests along with speeding them up (can create only the fixtures we need, when we need them).
- Started introducing new tests with pytest, but never rewrote the old ones.

I think that full rewrite will probably take too much effort, but a gradual rewrite with guidance of introducing new tests only with pytest makes sense to me.

**Reply by guidopetri** [reply]
That makes sense, I think a lot of projects follow a similar route. :D Personally, I don't mind spending some time rewriting the old tests. It's a good way for me to get to know the codebase and improve my testing at the same time.
